File: binary-search.cpp
// https://leetcode.com/problems/binary-search/

#include <iostream>
#include "bits/stdc++.h"

#define mid(a, b) ((a+b)/2)
using namespace std;

class Solution {
public:
	vector <int> arr;
	
	int binary_search(int left, int right, int target) {
		while (true) {
			if (left > right || right < left)return -1;
			int mid = mid(left, right);
			if (arr[mid] == target)return mid;
			else if (arr[mid] < target)left = mid + 1;
			else right = mid - 1;
			
		}
	}
	
	int search(vector <int> &nums, int target) {
		arr = nums;
		return binary_search(0, arr.size() - 1, target);
	}
};

File: binary-search.py
# https://leetcode.com/problems/binary-search/

def getMid(right, left):
    return (left + right) >> 1


class Solution:
    arr = []

    # it's usually implement recursive binarySearch. but iterative is much faster than recursive in python.
    def binarySearch(self, left, right, target):
        while 1:
            if left > right:
                return left
            mid = getMid(left, right)
            if self.arr[mid] == target:
                return mid
            elif self.arr[mid] < target:
                left = mid + 1
            else:
                right = mid - 1

    def search(self, nums: list[int], target: int) -> int:
        self.arr = nums
        return self.binarySearch(0, len(self.arr) - 1, target)


a = Solution()
print(a.search([1, 3, 5, 6], 0))
